---
output: github_document
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "80%",
  warning = FALSE,
  message = FALSE
)
```

# scib

<!-- badges: start -->
<!-- badges: end -->

The goal of scib (single-cell integration and benchmarking) is to provide a unified way to perform data integration and benchmarking when the ground truth is available.


# INSTALLATION

You can install the latest version scib like so:

```{r, eval=FALSE}
library(remotes)
install_github("mthimma/scib")
```

# LOAD DATA FOR DEMONSTRATION

We will load the `pbmc.demo` dataset that is used in the fundamental Seurat vignette. This is a Seurat object containing 2,700 peripheral blood mononuclear cells (PBMC) from 10X genomics. See `help(pbmc.demo)` for more details.

```{r}
library(tidyverse)
library(Seurat)

library(scib)
data(pbmc.demo)
pbmc.demo
```


## Quality control

Like all single-cell dataset, we need to perform quality control to ensure high quality dataset. Here we conduct a lenient quality control.

```{r}
pbmc.demo$percent.mt <- PercentageFeatureSet(pbmc.demo, pattern = "^MT-|^Mt-")

VlnPlot(pbmc.demo, c("nCount_RNA", "nFeature_RNA", "percent.mt"))

pbmc.demo <- subset(pbmc.demo,
                      nFeature_RNA > 200 &
                      nFeature_RNA < 2500 &
                      percent.mt   < 5)
```

Aside: After quality control

```{r}
pbmc.demo

VlnPlot(pbmc.demo, c("nCount_RNA", "nFeature_RNA", "percent.mt"))
```

## Ground truth

This dataset includes `seurat_annotations` provided by the Seurat co-authors. We also included a `batch` column which is artificially created for this package demonstration.

```{r}
pbmc.demo@meta.data %>%
  glimpse()
```

```{r}
pbmc.demo@meta.data %>% 
  janitor::tabyl(seurat_annotations, batch)
```

# DATA INTEGRATION

Aside: Seurat v5 stores information in the following layers

* `counts` layer: raw un-normalized counts
* `data` layer: normalized data, which is set correctly after `NormalizeData()`
* `scale.data` layer: z-scored/variance-stabilized data, which is set after `ScaleData()`

Here are the layers before the split

```{r}
Layers(pbmc.demo)
```


## Step 1: Split the Seurat object

We will split the RNA assay by the grouping factor that defines the batch effect. In reality, batch effects may arise from differences between donors, single-cell technology, sequencing platforms, timing, reagents, or experimental conditions across laboratories.

```{r}
pbmc.demo[["RNA"]] <- split(pbmc.demo[["RNA"]], f = pbmc.demo$batch)
```

Aside: We now can see that the counts and data layer has been split into three layers each.

```{r}
Layers(pbmc.demo)
```

## Step 2: Preprocess each split

Next we need to run the standard preprocessing steps for each split. This can be done via:

```{r}
pbmc.demo <- pbmc.demo %>%
  NormalizeData(verbose = FALSE) %>%
  FindVariableFeatures(verbose = FALSE) %>%
  ScaleData(verbose = FALSE) %>%
  RunPCA(verbose = FALSE)
```

We can see this has created the `pca` reduction based on the unintegrated data. We can determine the dimensional of this dataset from the Elbow plot to be approximately 10.

```{r}
Reductions(pbmc.demo)

ElbowPlot(pbmc.demo, ndims = 50, reduction = "pca")

ndims <- 10
```


## Step 3: Data integration step

This is the workhorse of the data integration process. Here is an example on how to execute this with Harmony. The Harmony-corrected embedding or integrated embedding is store in the `integrated.harmony` reduction.

```{r}
pbmc.demo <- IntegrateLayers(pbmc.demo,
                             method         = "CONOSIntegration",
                             orig.reduction = "pca",
                             new.reduction  = "integrated.conos",
                             verbose        = FALSE,
                             resolution     = 0.80,
                             ndims          = ndims)

Reductions(pbmc.demo)
```

Other available methods that can be used with `IntegrateLayers()`:


| method  | package             | Publication | 
|---      |---                  |---          |
| Harmony | Seurat              | ??          |
| CCA     | Seurat              | ??          |
| RPCA    | Seurat              | ??          |
| SCVI    | SeuratWrapper       | ??          |
| FastMNN | SeuratWrapper       | ??          |
| LIGER   | scib (this package) | ??          |
| BBKNN   | scib (this package) | ??          |



## Step 4: Cleanup after data integration

After the data integration, we no longer need the counts and data to be split by the batch variable. Therefore we can join the layers. 

```{r}
Layers(pbmc.demo)  # before

pbmc.demo <- JoinLayers(pbmc.demo)

Layers(pbmc.demo)  # after
```
The `scale.data` layer is a non-sparse matrix and not needed for further analysis. Therefore, we can remove this layer to reduce the object size.

```{r}
object.size(pbmc.demo) %>% format(units = "Mb")  # before

pbmc.demo[["RNA"]]$scale.data  <- NULL
pbmc.demo@assays$RNA@meta.data <- data.frame()

object.size(pbmc.demo) %>% format(units = "Mb")  # after
```


## Step 5: Processing the integrated data

We can now run the non-linear dimension reduction on the integrated embedding and save it. Here the `ndims` is the dimensionality of the dataset that we determined earlier.

```{r}
pbmc.demo <- RunUMAP(pbmc.demo,
                     dims           = 1:ndims,
                     reduction      = "integrated.conos",
                     reduction.name = "umap.conos",
                     verbose        = FALSE
) 

DimPlot(pbmc.demo, reduction = "umap.conos", group.by = "seurat_annotations", label = T) +
  NoLegend()
```

Alternatively, we can also cluster the cells at certain resolution based on the integrated embedding.

```{r}
pbmc.demo <- FindNeighbors(pbmc.demo, reduction = "integrated.conos")

pbmc.demo <- FindClusters (pbmc.demo, res = 0.80)

DimPlot(pbmc.demo,
        reduction = "umap.conos",
        group.by  = "RNA_snn_res.0.8",
        label     = TRUE) +
  NoLegend()
```


# EVALUATION

## Cross tabulation with the ground truth

We can compare the clusters based on the integrated embedding vs. the ground truth provided by the Seurat authors:

```{r}
tb <- table(pbmc.demo$seurat_annotations, pbmc.demo$RNA_snn_res.0.8)

print.table(tb, zero.print = ".")
```

There are several ways of estimating the performance of the data integration technique. Accuracy is one of the simpler method. 

1. Assign every cluster to one label using simple majority (e.g. 0 = Naive CD4 T, 1 = Memory CD4 T, 2 = B, ..., 11 = Platelet). 

2. Add up the numbers that correspond to this majority based assignment (384 + 337 + 343 + 234 + 241 + 232 + 129 + 152 + 148 + 130 + 32 + 14). This totals to 2,376 here.

3. Divide by the total number of cells. i.e. 2376 / 2638 which gives 90%.

However, there are more sophisticated metrics for evaluation which we cover next.


## Evaluate the data integration

```{r, include=FALSE}
# save(pbmc.demo, file = "data/pbmc.demo_final.rda")

library(tidyverse)
library(Seurat)
library(scib)
load("data/pbmc.demo_final.rda")
```


```{r}
results <- run_eval_metrics(pbmc.demo, 
                            reduction      = "integrated.conos", 
                            groundtruth.cn = "seurat_annotations", 
                            predicted.cn   = "RNA_snn_res.0.8")

data.frame(results)
```


# EVALUATING MULTIPLE METHODS

Let us demonstrate how to run and evaluate mutiple data integration techniques

```{r}
# Reset the data
data(pbmc.demo)

# Step 1
pbmc.demo[["RNA"]] <- split(pbmc.demo[["RNA"]], f = pbmc.demo$batch)

# Step 2
pbmc.demo <- pbmc.demo %>%
  NormalizeData(verbose = FALSE) %>%
  FindVariableFeatures(verbose = FALSE) %>%
  ScaleData(verbose = FALSE) %>%
  RunPCA(verbose = FALSE)

ndims <- 10  # determined from Elbowplot(pbmc.demo)
```


```{r}
methods <- c(harmony = "HarmonyIntegration",
             conos   = "CONOSIntegration")

for(i in 1:length(methods)){
  
  pbmc.demo <- IntegrateLayers(pbmc.demo,
                               method         = methods[i],
                               orig.reduction = "pca",
                               new.reduction  = paste0("integrated.", names(methods)[i]),
                               verbose        = FALSE,
                               resolution     = 0.80,
                               ndims          = ndims)
  
  cleanup 
  
  postprocessing
}

```
